src/_pytest/terminal.py

```diff
@@ -233,29 +238,30 @@ def get_location(self, config):
 
 
 class TerminalReporter:
-    def __init__(self, config, file=None):
+    def __init__(self, config: Config, file=None):
         import _pytest.config
 
         self.config = config
         self._numcollected = 0
-        self._session = None
+        self._session = None  # type: Optional[Session]
         self._showfspath = None
 
-        self.stats = {}
+        # XXX: needs TypedDict?  warnings => WarningReport, others => BaseReport
```

Comment:
Ref: https://www.python.org/dev/peps/pep-0589/

Reply:
Can do it, only issue is that unless we want to bring in a *runtime* dependency on `typing_extension`, we'll need to add a compat for it, or always surround it with `if TYPE_CHECKING` and string annotations.

Reply:
I think it is rather out of scope, since it requires py38 (IIRC).
Better options?
Otherwise I will remove the XXX/TODO comments about it.

We could also have `stats_warnings` explicitly for this, but that's not B/C (but could be done via a property or thelike then).
But IIRC I've tried a property based approach (where I would define the returned type), which did not help really.